### PR TITLE
Line-up snapshot listeners resolution to avoid double calls

### PR DESCRIPTION
**Pull request by volodk85**
Snapshot listeners can be concurrently resolved from two different *clusterApplierService* and *masterService* task threads. If a listener is a **single action** listener, meaning that it has to be resolved only once, the following traces can occur, see #101876 
```
at org.elasticsearch.action.ActionListener$4.assertFirstRun(ActionListener.java:324)
at org.elasticsearch.action.ActionListener$4.onFailure(ActionListener.java:335)
```

Fix

Line up *resolve listener and remove it from tracking collection* operations over snapshot listeners in order to avoid double invocation by separate threads

Fix for #101876

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Makes sense to me, but this highlights (and slightly exacerbates) a problem with how we're synchronizing on `currentlyFinalizing`: we shouldn't be completing listeners while holding this lock. Those listeners could be doing anything, and if they take a while then that holds up other waiting threads. Could you collect up the work to do (e.g. in a `List<Runnable>`) under the mutex but delay its execution until the mutex is released?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -1897,11 +1913,6 @@ private static SnapshotDeletionsInProgress deletionsWithoutSnapshots(
         return changed ? SnapshotDeletionsInProgress.of(updatedEntries) : null;
     }
 
-    private void failSnapshotCompletionListeners(Snapshot snapshot, Exception e) {
```
Comment (volodk85) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1900]: removed `failSnapshotCompletionListeners ` method as it is a "collection of side-effects". inlined its logic for better readability

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1900]: I'd prefer to keep this in its own method, it's used in a lot of different places that now have duplicated code which I expect will drift apart over time now.

**Review by DaveCTurner — COMMENTED** [review]
The inlinings and other refactorings make this pretty hard to review, and I'm not convinced they improve the code here either. Could we separate them out into two PRs, one which fixes the bug without making too many other changes, and another which we can review separately about the code structure?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (volodk85) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I see, brought method back and overloaded it with extra param - failing listeners collector

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 2]
```diff
@@ -2466,17 +2481,22 @@ protected void handleListeners(List<ActionListener<Void>> deleteListeners) {
...
-                for (Snapshot snapshot : Set.copyOf(snapshotCompletionListeners.keySet())) {
-                    failSnapshotCompletionListeners(snapshot, new SnapshotException(snapshot, "no longer master"));
+                for (final Snapshot snapshot : snapshotCompletionListeners.keySet()) {
+                    failSnapshotCompletionListeners(
+                        snapshot,
+                        new SnapshotException(snapshot, "no longer master"),
+                        readyToResolveListeners::add
+                    );
                 }
                 final Exception wrapped = new RepositoryException("_all", "Failed to update cluster state during repository operation", e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2495]: We should probably guard this one against unnecessary creation too - unlike the one in my previous comment this one is actually expensive (it computes the stack trace). That can be a follow-up tho, keeping this PR focussed just on the one change.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 3]
```diff
@@ -1902,6 +1911,12 @@ private void failSnapshotCompletionListeners(Snapshot snapshot, Exception e) {
         assert repositoryOperations.assertNotQueued(snapshot);
     }
 
+    private void failSnapshotCompletionListeners(Snapshot snapshot, Exception e, Consumer<Runnable> failingListenersConsumer) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1914]: Nit: maybe reimplement the other `failSnapshotCompletionListeners` as `failSnapshotCompletionListeners(snapshot, e, Runnable::run)`. And possibly just inline it, adding the `Runnable::run` parameter to its callers.

**Review by DaveCTurner — APPROVED** [review]
Looks good; I left a few (optional) nits.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 2 above)
Comment (volodk85) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2498]: I'll do this as a follow-up

**Pull request merged**